### PR TITLE
Add support for ApiKey V3 in authentication flow

### PR DESCRIPTION
**Pull request by skofman1**
If we meet a V3  ApiKey in the DB, we will be able to authenticate with it.
Actually hashing V1/V2 ApiKeys into V3 will be in a follow-up PR.

### CODE REVIEW

File: tests/NuGetGallery.Facts/Infrastructure/Authentication/ApiKeyV3Facts.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+            // Act & Assert
+            Assert.True(parsedApiKeyV3.Verify(createdApiKeyV3.HashedApiKey));
+        }
+    }
+
+
+
+
+
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Infrastructure/Authentication/ApiKeyV3Facts.cs]: Maybe a bit too much whitespace

File: src/NuGetGallery/Infrastructure/Authentication/CredentialValidator.cs [hunk 2]
```diff
@@ -43,8 +43,29 @@ public IList<Credential> GetValidCredentialsForApiKey(IQueryable<Credential> all
...
+                        if (credential.Value == providedApiKey)
+                        {
+                            return credential.Type == CredentialTypes.ApiKey.V1 ||
+                                   credential.Type == CredentialTypes.ApiKey.V2 ||
+                                   credential.Type == CredentialTypes.ApiKey.VerifyV1;
+                        }
+                        else if (credential.Type == CredentialTypes.ApiKey.V3)
+                        {
+                            return v3ApiKey.Verify(credential.Value);
+                        }
```
Comment (chenriksson) [inline src/NuGetGallery/Infrastructure/Authentication/CredentialValidator.cs]: nit: maybe switch is appropriate here?
```
switch (credential.Type) {
    case CredentialTypes.ApiKey.V1:
    case CredentialTypes.ApiKey.V2:
    case CredentialTypes.ApiKey.VerifyV1:
        return credential.Value == providedApiKey;
    case CredentialTypes.ApiKey.V3:
        return v3ApiKey.Verify(credential.Value);
    default:
        return false;
}
```

Comment (skofman1) [reply src/NuGetGallery/Infrastructure/Authentication/CredentialValidator.cs]: done

File: src/NuGetGallery/Infrastructure/Authentication/ApiKeyV3.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        private ApiKeyV3()
+        {
+        }
+
+        /// <summary>
+        /// Creates an ApiKeyV3 from an APIKey V1/V2 format (GUID).
+        /// </summary>
+        public static ApiKeyV3 CreateFromV1V2ApiKey(string plaintextApiKey)
+        {
+            // Since V1/V2/V3 have the same formar (Guid), we can use the same parse method
```
Comment (shishirx34) [inline src/NuGetGallery/Infrastructure/Authentication/ApiKeyV3.cs]: nit: typo format

File: src/NuGetGallery/Infrastructure/Authentication/CredentialValidator.cs [hunk 4]
```diff
@@ -43,8 +43,34 @@ public IList<Credential> GetValidCredentialsForApiKey(IQueryable<Credential> all
...
+                                }
+                            case CredentialTypes.ApiKey.V3:
+                                {
+                                    return v3ApiKey.Verify(credential.Value);
+                                }
+
+                            default:
+                                {
+                                    return false;
+                                }
```
Comment (shishirx34) [inline src/NuGetGallery/Infrastructure/Authentication/CredentialValidator.cs]: nit: no need for braces in `case:`

Comment (skofman1) [reply src/NuGetGallery/Infrastructure/Authentication/CredentialValidator.cs]: I like braces in case.. that's the pattern I always use..

**Pull request merged**